{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9d562681-9772-4edc-84bb-469052f3d93d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import ipywidgets as widgets\n",
    "import plotly.express as px\n",
    "from IPython.display import clear_output\n",
    "\n",
    "import dask.dataframe as dd\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "read_csv_dtype = {'fasta': \"object\",\n",
    " 'rank': \"object\",\n",
    " 'ko_id': \"object\",\n",
    " 'kegg_hit': \"object\",\n",
    " 'peptidase_id': \"object\",\n",
    " 'peptidase_family': \"object\",\n",
    " 'peptidase_hit': \"object\",\n",
    " 'peptidase_RBH': \"object\",\n",
    " 'peptidase_identity': \"float64\",\n",
    " 'peptidase_bitScore': \"float64\",\n",
    " 'peptidase_eVal': \"float64\",\n",
    " 'pfam_hits': \"object\",\n",
    " 'cazy_ids': \"object\",\n",
    " 'cazy_hits': \"object\",\n",
    " 'cazy_subfam_ec': \"object\",\n",
    " 'cazy_best_hit': \"object\",\n",
    " 'vogdb_id': \"object\",\n",
    " 'vogdb_hits': \"object\",\n",
    " 'vogdb_categories': \"object\",\n",
    " 'heme_regulatory_motif_count': \"float64\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c4aab76-f03d-435a-8c93-724af9ad3e40",
   "metadata": {},
   "source": [
    "Traditionally, methanogenesis and archaeal methanotrophy restricted to eight orders within Euryarchaeota:\n",
    "- Methanopyrales\n",
    "- Methanobacteriales\n",
    "- Methanosarcinales\n",
    "- Methanomassiliicoccales\n",
    "- Candidatus Methanophagales (ANME-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4336582c-ef66-4281-953e-fbba778bb78e",
   "metadata": {},
   "source": [
    "![](https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41579-018-0136-7/MediaObjects/41579_2018_136_Fig2_HTML.png?as=webp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84fe4f7b-0d99-474f-9854-94598cb99534",
   "metadata": {},
   "source": [
    "This is the Archaeal genome tree.\n",
    "\n",
    "The blue ones indicates hydrogenotrophic/aceticlastic/methylotrophic Euryarchaeota:\n",
    "- Methanopyrus kandleri\n",
    "- Methanobacteriales\n",
    "- Methanococcales\n",
    "- Methanomicrobiales\n",
    "- Ca. Methanoflorens stordalenmirensis\n",
    "- Methanocellales\n",
    "- Methanosarcinales\n",
    "    - Methermicoccaceae\n",
    "    - Methanosaetaceae\n",
    "    - Methanosarcinaceae\n",
    "    - Ca. Methanoperedens\n",
    "- Ca. Methanophagales\n",
    "- Ca. Syntrophoarchaeum\n",
    "\n",
    "Yellow is methanotrophic Euryarchaeota:\n",
    "- Ca. Methanoperedens\n",
    "- Ca. Methanophagales\n",
    "\n",
    "Red is alkanotrophs:\n",
    "- Ca. Bathyarchaeota (TACK)\n",
    "- Ca. Syntrophoarchaeum (Euryarchaeota)\n",
    "\n",
    "Green is H2-dependent methylotrophic:\n",
    "- Ca. Vestrearchaeota (TACK)\n",
    "- Ca. Methanofastidiosa\n",
    "- Methanonatronarchaeia\n",
    "- Methanomassiliicoccales"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "046d75d8-74e4-4730-b248-7538d12ce745",
   "metadata": {},
   "source": [
    "Who has mcrA [KEGG: K00399] in my genomes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5c95c85a-b2fd-4679-8e76-868078da5fe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_all = pd.read_csv(\"../../input_folder/summary_all.tsv\", sep=\"\\t\", index_col=0)\n",
    "summary_hq = pd.read_csv(\"../../input_folder/summary_hq.tsv\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "genomes = summary_all.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "283bb998-9798-4af1-8442-c53356143e06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Contamination</th>\n",
       "      <th>Contig_N50</th>\n",
       "      <th>GC_Content</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>average_coverage</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>16S_rRNA</th>\n",
       "      <th>23S_rRNA</th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>classification</th>\n",
       "      <th>red_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1C3_F_metabat.43</th>\n",
       "      <td>96.65</td>\n",
       "      <td>2.03</td>\n",
       "      <td>98027</td>\n",
       "      <td>0.50</td>\n",
       "      <td>2863110</td>\n",
       "      <td>13.75</td>\n",
       "      <td>67</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...</td>\n",
       "      <td>0.93561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2E1_F_metabat.953</th>\n",
       "      <td>93.49</td>\n",
       "      <td>0.08</td>\n",
       "      <td>451638</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2142996</td>\n",
       "      <td>15.33</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...</td>\n",
       "      <td>0.93583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1E5_M_semibin.11763</th>\n",
       "      <td>92.27</td>\n",
       "      <td>0.39</td>\n",
       "      <td>1841381</td>\n",
       "      <td>0.44</td>\n",
       "      <td>2607823</td>\n",
       "      <td>36.52</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...</td>\n",
       "      <td>0.74397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2E1_M_semibin.1716</th>\n",
       "      <td>90.52</td>\n",
       "      <td>2.37</td>\n",
       "      <td>1432256</td>\n",
       "      <td>0.45</td>\n",
       "      <td>2862093</td>\n",
       "      <td>156.84</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...</td>\n",
       "      <td>0.74395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3C4_F_semibin.3639</th>\n",
       "      <td>97.36</td>\n",
       "      <td>1.08</td>\n",
       "      <td>2729341</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3054849</td>\n",
       "      <td>18.27</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...</td>\n",
       "      <td>0.74317</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Completeness  Contamination  Contig_N50  GC_Content  \\\n",
       "C1C3_F_metabat.43            96.65           2.03       98027        0.50   \n",
       "C2E1_F_metabat.953           93.49           0.08      451638        0.52   \n",
       "C1E5_M_semibin.11763         92.27           0.39     1841381        0.44   \n",
       "C2E1_M_semibin.1716          90.52           2.37     1432256        0.45   \n",
       "C3C4_F_semibin.3639          97.36           1.08     2729341        0.44   \n",
       "\n",
       "                      genome_size  average_coverage  number_of_contigs  \\\n",
       "C1C3_F_metabat.43         2863110             13.75                 67   \n",
       "C2E1_F_metabat.953        2142996             15.33                  7   \n",
       "C1E5_M_semibin.11763      2607823             36.52                  4   \n",
       "C2E1_M_semibin.1716       2862093            156.84                  4   \n",
       "C3C4_F_semibin.3639       3054849             18.27                 17   \n",
       "\n",
       "                      16S_rRNA  23S_rRNA  5S_rRNA  \\\n",
       "C1C3_F_metabat.43         True     False    False   \n",
       "C2E1_F_metabat.953        True      True     True   \n",
       "C1E5_M_semibin.11763      True      True     True   \n",
       "C2E1_M_semibin.1716       True      True     True   \n",
       "C3C4_F_semibin.3639       True      True     True   \n",
       "\n",
       "                                                         classification  \\\n",
       "C1C3_F_metabat.43     d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...   \n",
       "C2E1_F_metabat.953    d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...   \n",
       "C1E5_M_semibin.11763  d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...   \n",
       "C2E1_M_semibin.1716   d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...   \n",
       "C3C4_F_semibin.3639   d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-...   \n",
       "\n",
       "                      red_value  \n",
       "C1C3_F_metabat.43       0.93561  \n",
       "C2E1_F_metabat.953      0.93583  \n",
       "C1E5_M_semibin.11763    0.74397  \n",
       "C2E1_M_semibin.1716     0.74395  \n",
       "C3C4_F_semibin.3639     0.74317  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_hq.head()"
   ]
  },
  {
   "cell_type": "raw",
   "id": "639a75c2-3396-42a1-8d5e-035669531aa2",
   "metadata": {},
   "source": [
    "def does_my_genome_has_this_kegg_gene(genome, kegg_gene):\n",
    "    df = dd.read_csv(f\"dram_faa/{genome}/annotations.tsv\", sep=\"\\t\", dtype=read_csv_dtype)\n",
    "    return not df[df[\"ko_id\"] == kegg_gene].compute().empty\n",
    "\n",
    "does_my_genome_has_this_kegg_gene(genomes[0], \"K00399\")"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c6a5cf28-6431-43c8-99e7-fd90a73824b9",
   "metadata": {},
   "source": [
    "mcrA_genomes = [\n",
    "    g for g in tqdm(genomes, desc=\"Filtering genomes for K00399\")\n",
    "    if does_my_genome_has_this_kegg_gene(g, \"K00399\")\n",
    "]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "69662c8b-8cc6-4f5c-8772-13ea8a2f1e0c",
   "metadata": {},
   "source": [
    "import subprocess\n",
    "\n",
    "def does_my_genome_have_this_kegg_gene_with_grep(genome, kegg_gene):\n",
    "    # Construct the file path\n",
    "    file_path = f\"dram_faa/{genome}/annotations.tsv\"\n",
    "    \n",
    "    # Use subprocess to run grep and capture output\n",
    "    result = subprocess.run(\n",
    "        [\"grep\", \"-q\", kegg_gene, file_path],\n",
    "        stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "    )\n",
    "    \n",
    "    # Check if grep found the kegg_gene\n",
    "    return result.returncode == 0  # grep returns 0 if found, non-zero if not\n",
    "\n",
    "# Example usage\n",
    "result = does_my_genome_have_this_kegg_gene_with_grep(genomes[0], \"K00399\")\n",
    "print(result)  # True if the KEGG gene exists, False otherwise"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6dffa81b-de5f-4b9c-bd43-12a09a5f7fd5",
   "metadata": {},
   "source": [
    "mcrA_genomes = [\n",
    "    g for g in tqdm(genomes, desc=\"Filtering genomes for K00399\")\n",
    "    if does_my_genome_have_this_kegg_gene_with_grep(g, \"K00399\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8eb28348-5c05-44bf-b53d-713b1bfab420",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "def does_my_genome_have_this_kegg_gene_with_grep2(genome, kegg_gene):\n",
    "    # Construct the file path\n",
    "    file_path = f\"../../input_folder/dram_faa/{genome}/annotations.tsv\"\n",
    "    \n",
    "    # Use subprocess to run grep and capture output, looking for exact match in the 4th column\n",
    "    result = subprocess.run(\n",
    "        [\"grep\", \"-P\", \"^([^\\t]*\\t){3}\" + kegg_gene, file_path],\n",
    "        stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "    )\n",
    "    \n",
    "    # If grep found the gene, return True, otherwise False\n",
    "    return result.returncode == 0  # grep returns 0 if found, non-zero if not\n",
    "\n",
    "# Example usage\n",
    "result = does_my_genome_have_this_kegg_gene_with_grep2(genomes[0], \"K00399\")\n",
    "print(result)  # True if the KEGG gene exists in the 4th column, False otherwise\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e13d0e97-e556-4001-b92e-93e0d9f450ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "should_be_true = \"C1E5_M_semibin.11763\"\n",
    "result = does_my_genome_have_this_kegg_gene_with_grep2(should_be_true, \"K00399\")\n",
    "print(result)  # True if the KEGG gene exists in the 4th column, False otherwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "39e05692-fc28-491c-81c1-00c0852226bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_all = pd.read_csv(\"../../input_folder/summary_all.tsv\", sep=\"\\t\", index_col=0)\n",
    "summary_hq = pd.read_csv(\"../../input_folder/summary_hq.tsv\", sep=\"\\t\", index_col=0)\n",
    "genomes = summary_all.index.tolist()\n",
    "# genomes=[\"C1E5_M_semibin.11763\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1fa38874-2417-4c8c-8f71-739076978fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c30f83c2a1fe4d2f9348198bc1a93611",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Filtering genomes for K00399:   0%|          | 0/2634 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from concurrent.futures import ProcessPoolExecutor, as_completed\n",
    "\n",
    "def filter_genomes_for_kegg_gene(genomes, gene, max_workers=6):\n",
    "    \"\"\"Filter genomes to check if they contain a specific KEGG gene using multiprocessing.\"\"\"\n",
    "    filtered_genomes = []\n",
    "    \n",
    "    # Use ProcessPoolExecutor for multiprocessing\n",
    "    with ProcessPoolExecutor(max_workers=max_workers) as executor:\n",
    "        # Submit tasks to the pool\n",
    "        futures = {\n",
    "            executor.submit(does_my_genome_have_this_kegg_gene_with_grep2, g, gene): g\n",
    "            for g in genomes\n",
    "        }\n",
    "\n",
    "        # Use tqdm to track progress\n",
    "        for future in tqdm(as_completed(futures), total=len(futures), desc=f\"Filtering genomes for {gene}\"):\n",
    "            genome = futures[future]\n",
    "            try:\n",
    "                # Check the result of each task\n",
    "                if future.result():\n",
    "                    filtered_genomes.append(genome)\n",
    "            except Exception as e:\n",
    "                print(f\"Error processing genome {genome}: {e}\")\n",
    "\n",
    "    return filtered_genomes\n",
    "\n",
    "# Example usage\n",
    "mcrA_genomes = filter_genomes_for_kegg_gene(genomes, \"K00399\", max_workers=6)\n",
    "\n",
    "# mcrA_genomes = [\n",
    "#     g for g in tqdm(genomes, desc=\"Filtering genomes for K00399\")\n",
    "#     if does_my_genome_have_this_kegg_gene_with_grep2(g, \"K00399\")\n",
    "# ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20b506ab-d9bc-4a90-a809-09d1c7f5a333",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_cdcd0\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_cdcd0_level0_col0\" class=\"col_heading level0 col0\" >Completeness</th>\n",
       "      <th id=\"T_cdcd0_level0_col1\" class=\"col_heading level0 col1\" >Contamination</th>\n",
       "      <th id=\"T_cdcd0_level0_col2\" class=\"col_heading level0 col2\" >Contig_N50</th>\n",
       "      <th id=\"T_cdcd0_level0_col3\" class=\"col_heading level0 col3\" >GC_Content</th>\n",
       "      <th id=\"T_cdcd0_level0_col4\" class=\"col_heading level0 col4\" >genome_size</th>\n",
       "      <th id=\"T_cdcd0_level0_col5\" class=\"col_heading level0 col5\" >average_coverage</th>\n",
       "      <th id=\"T_cdcd0_level0_col6\" class=\"col_heading level0 col6\" >number_of_contigs</th>\n",
       "      <th id=\"T_cdcd0_level0_col7\" class=\"col_heading level0 col7\" >16S_rRNA</th>\n",
       "      <th id=\"T_cdcd0_level0_col8\" class=\"col_heading level0 col8\" >23S_rRNA</th>\n",
       "      <th id=\"T_cdcd0_level0_col9\" class=\"col_heading level0 col9\" >5S_rRNA</th>\n",
       "      <th id=\"T_cdcd0_level0_col10\" class=\"col_heading level0 col10\" >classification</th>\n",
       "      <th id=\"T_cdcd0_level0_col11\" class=\"col_heading level0 col11\" >red_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row0\" class=\"row_heading level0 row0\" >C2E1_F_metabat.953</th>\n",
       "      <td id=\"T_cdcd0_row0_col0\" class=\"data row0 col0\" >93.490000</td>\n",
       "      <td id=\"T_cdcd0_row0_col1\" class=\"data row0 col1\" >0.080000</td>\n",
       "      <td id=\"T_cdcd0_row0_col2\" class=\"data row0 col2\" >451638</td>\n",
       "      <td id=\"T_cdcd0_row0_col3\" class=\"data row0 col3\" >0.520000</td>\n",
       "      <td id=\"T_cdcd0_row0_col4\" class=\"data row0 col4\" >2142996</td>\n",
       "      <td id=\"T_cdcd0_row0_col5\" class=\"data row0 col5\" >15.330000</td>\n",
       "      <td id=\"T_cdcd0_row0_col6\" class=\"data row0 col6\" >7</td>\n",
       "      <td id=\"T_cdcd0_row0_col7\" class=\"data row0 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row0_col8\" class=\"data row0 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row0_col9\" class=\"data row0 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row0_col10\" class=\"data row0 col10\" >d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-38;f__Bog-38;g__Bog-38;s__</td>\n",
       "      <td id=\"T_cdcd0_row0_col11\" class=\"data row0 col11\" >0.935830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row1\" class=\"row_heading level0 row1\" >C1E5_M_semibin.11763</th>\n",
       "      <td id=\"T_cdcd0_row1_col0\" class=\"data row1 col0\" >92.270000</td>\n",
       "      <td id=\"T_cdcd0_row1_col1\" class=\"data row1 col1\" >0.390000</td>\n",
       "      <td id=\"T_cdcd0_row1_col2\" class=\"data row1 col2\" >1841381</td>\n",
       "      <td id=\"T_cdcd0_row1_col3\" class=\"data row1 col3\" >0.440000</td>\n",
       "      <td id=\"T_cdcd0_row1_col4\" class=\"data row1 col4\" >2607823</td>\n",
       "      <td id=\"T_cdcd0_row1_col5\" class=\"data row1 col5\" >36.520000</td>\n",
       "      <td id=\"T_cdcd0_row1_col6\" class=\"data row1 col6\" >4</td>\n",
       "      <td id=\"T_cdcd0_row1_col7\" class=\"data row1 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row1_col8\" class=\"data row1 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row1_col9\" class=\"data row1 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row1_col10\" class=\"data row1 col10\" >d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-38;f__Bog-38;g__;s__</td>\n",
       "      <td id=\"T_cdcd0_row1_col11\" class=\"data row1 col11\" >0.743970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row2\" class=\"row_heading level0 row2\" >C2E1_M_semibin.1716</th>\n",
       "      <td id=\"T_cdcd0_row2_col0\" class=\"data row2 col0\" >90.520000</td>\n",
       "      <td id=\"T_cdcd0_row2_col1\" class=\"data row2 col1\" >2.370000</td>\n",
       "      <td id=\"T_cdcd0_row2_col2\" class=\"data row2 col2\" >1432256</td>\n",
       "      <td id=\"T_cdcd0_row2_col3\" class=\"data row2 col3\" >0.450000</td>\n",
       "      <td id=\"T_cdcd0_row2_col4\" class=\"data row2 col4\" >2862093</td>\n",
       "      <td id=\"T_cdcd0_row2_col5\" class=\"data row2 col5\" >156.840000</td>\n",
       "      <td id=\"T_cdcd0_row2_col6\" class=\"data row2 col6\" >4</td>\n",
       "      <td id=\"T_cdcd0_row2_col7\" class=\"data row2 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row2_col8\" class=\"data row2 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row2_col9\" class=\"data row2 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row2_col10\" class=\"data row2 col10\" >d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-38;f__Bog-38;g__;s__</td>\n",
       "      <td id=\"T_cdcd0_row2_col11\" class=\"data row2 col11\" >0.743950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row3\" class=\"row_heading level0 row3\" >C7F4_F_metabat.1057</th>\n",
       "      <td id=\"T_cdcd0_row3_col0\" class=\"data row3 col0\" >94.390000</td>\n",
       "      <td id=\"T_cdcd0_row3_col1\" class=\"data row3 col1\" >1.360000</td>\n",
       "      <td id=\"T_cdcd0_row3_col2\" class=\"data row3 col2\" >1606116</td>\n",
       "      <td id=\"T_cdcd0_row3_col3\" class=\"data row3 col3\" >0.430000</td>\n",
       "      <td id=\"T_cdcd0_row3_col4\" class=\"data row3 col4\" >2651675</td>\n",
       "      <td id=\"T_cdcd0_row3_col5\" class=\"data row3 col5\" >26.030000</td>\n",
       "      <td id=\"T_cdcd0_row3_col6\" class=\"data row3 col6\" >2</td>\n",
       "      <td id=\"T_cdcd0_row3_col7\" class=\"data row3 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row3_col8\" class=\"data row3 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row3_col9\" class=\"data row3 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row3_col10\" class=\"data row3 col10\" >d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-38;f__Bog-38;g__;s__</td>\n",
       "      <td id=\"T_cdcd0_row3_col11\" class=\"data row3 col11\" >0.744810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row4\" class=\"row_heading level0 row4\" >C7F4_M_semibin.4823</th>\n",
       "      <td id=\"T_cdcd0_row4_col0\" class=\"data row4 col0\" >96.250000</td>\n",
       "      <td id=\"T_cdcd0_row4_col1\" class=\"data row4 col1\" >0.480000</td>\n",
       "      <td id=\"T_cdcd0_row4_col2\" class=\"data row4 col2\" >729249</td>\n",
       "      <td id=\"T_cdcd0_row4_col3\" class=\"data row4 col3\" >0.440000</td>\n",
       "      <td id=\"T_cdcd0_row4_col4\" class=\"data row4 col4\" >2922066</td>\n",
       "      <td id=\"T_cdcd0_row4_col5\" class=\"data row4 col5\" >38.100000</td>\n",
       "      <td id=\"T_cdcd0_row4_col6\" class=\"data row4 col6\" >8</td>\n",
       "      <td id=\"T_cdcd0_row4_col7\" class=\"data row4 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row4_col8\" class=\"data row4 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row4_col9\" class=\"data row4 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row4_col10\" class=\"data row4 col10\" >d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-38;f__Bog-38;g__;s__</td>\n",
       "      <td id=\"T_cdcd0_row4_col11\" class=\"data row4 col11\" >0.740590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row5\" class=\"row_heading level0 row5\" >C3C4_F_semibin.3639</th>\n",
       "      <td id=\"T_cdcd0_row5_col0\" class=\"data row5 col0\" >97.360000</td>\n",
       "      <td id=\"T_cdcd0_row5_col1\" class=\"data row5 col1\" >1.080000</td>\n",
       "      <td id=\"T_cdcd0_row5_col2\" class=\"data row5 col2\" >2729341</td>\n",
       "      <td id=\"T_cdcd0_row5_col3\" class=\"data row5 col3\" >0.440000</td>\n",
       "      <td id=\"T_cdcd0_row5_col4\" class=\"data row5 col4\" >3054849</td>\n",
       "      <td id=\"T_cdcd0_row5_col5\" class=\"data row5 col5\" >18.270000</td>\n",
       "      <td id=\"T_cdcd0_row5_col6\" class=\"data row5 col6\" >17</td>\n",
       "      <td id=\"T_cdcd0_row5_col7\" class=\"data row5 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row5_col8\" class=\"data row5 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row5_col9\" class=\"data row5 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row5_col10\" class=\"data row5 col10\" >d__Archaea;p__Halobacteriota;c__Bog-38;o__Bog-38;f__Bog-38;g__;s__</td>\n",
       "      <td id=\"T_cdcd0_row5_col11\" class=\"data row5 col11\" >0.743170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row6\" class=\"row_heading level0 row6\" >C3D5_M_semibin.1409</th>\n",
       "      <td id=\"T_cdcd0_row6_col0\" class=\"data row6 col0\" >94.300000</td>\n",
       "      <td id=\"T_cdcd0_row6_col1\" class=\"data row6 col1\" >0.670000</td>\n",
       "      <td id=\"T_cdcd0_row6_col2\" class=\"data row6 col2\" >221283</td>\n",
       "      <td id=\"T_cdcd0_row6_col3\" class=\"data row6 col3\" >0.620000</td>\n",
       "      <td id=\"T_cdcd0_row6_col4\" class=\"data row6 col4\" >2141553</td>\n",
       "      <td id=\"T_cdcd0_row6_col5\" class=\"data row6 col5\" >6.660000</td>\n",
       "      <td id=\"T_cdcd0_row6_col6\" class=\"data row6 col6\" >26</td>\n",
       "      <td id=\"T_cdcd0_row6_col7\" class=\"data row6 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row6_col8\" class=\"data row6 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row6_col9\" class=\"data row6 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row6_col10\" class=\"data row6 col10\" >d__Archaea;p__Halobacteriota;c__Methanomicrobia;o__Methanomicrobiales;f__JACTUA01;g__JACTUA01;s__</td>\n",
       "      <td id=\"T_cdcd0_row6_col11\" class=\"data row6 col11\" >0.957050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row7\" class=\"row_heading level0 row7\" >C2E1_M_semibin.27400</th>\n",
       "      <td id=\"T_cdcd0_row7_col0\" class=\"data row7 col0\" >94.470000</td>\n",
       "      <td id=\"T_cdcd0_row7_col1\" class=\"data row7 col1\" >0.810000</td>\n",
       "      <td id=\"T_cdcd0_row7_col2\" class=\"data row7 col2\" >2074696</td>\n",
       "      <td id=\"T_cdcd0_row7_col3\" class=\"data row7 col3\" >0.600000</td>\n",
       "      <td id=\"T_cdcd0_row7_col4\" class=\"data row7 col4\" >2074698</td>\n",
       "      <td id=\"T_cdcd0_row7_col5\" class=\"data row7 col5\" >32.590000</td>\n",
       "      <td id=\"T_cdcd0_row7_col6\" class=\"data row7 col6\" >1</td>\n",
       "      <td id=\"T_cdcd0_row7_col7\" class=\"data row7 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row7_col8\" class=\"data row7 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row7_col9\" class=\"data row7 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row7_col10\" class=\"data row7 col10\" >d__Archaea;p__Halobacteriota;c__Methanomicrobia;o__Methanomicrobiales;f__JACTUA01;g__JACTUA01;s__</td>\n",
       "      <td id=\"T_cdcd0_row7_col11\" class=\"data row7 col11\" >0.956900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row8\" class=\"row_heading level0 row8\" >C2E1_P_metabat.289</th>\n",
       "      <td id=\"T_cdcd0_row8_col0\" class=\"data row8 col0\" >99.430000</td>\n",
       "      <td id=\"T_cdcd0_row8_col1\" class=\"data row8 col1\" >2.060000</td>\n",
       "      <td id=\"T_cdcd0_row8_col2\" class=\"data row8 col2\" >2242585</td>\n",
       "      <td id=\"T_cdcd0_row8_col3\" class=\"data row8 col3\" >0.450000</td>\n",
       "      <td id=\"T_cdcd0_row8_col4\" class=\"data row8 col4\" >2242705</td>\n",
       "      <td id=\"T_cdcd0_row8_col5\" class=\"data row8 col5\" >13.780000</td>\n",
       "      <td id=\"T_cdcd0_row8_col6\" class=\"data row8 col6\" >1</td>\n",
       "      <td id=\"T_cdcd0_row8_col7\" class=\"data row8 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row8_col8\" class=\"data row8 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row8_col9\" class=\"data row8 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row8_col10\" class=\"data row8 col10\" >d__Archaea;p__Halobacteriota;c__Syntropharchaeia;o__Alkanophagales;f__Methanospirareceae;g__ANME-1-THS;s__</td>\n",
       "      <td id=\"T_cdcd0_row8_col11\" class=\"data row8 col11\" >0.947890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row9\" class=\"row_heading level0 row9\" >C2D1_F_semibin.3899</th>\n",
       "      <td id=\"T_cdcd0_row9_col0\" class=\"data row9 col0\" >100.000000</td>\n",
       "      <td id=\"T_cdcd0_row9_col1\" class=\"data row9 col1\" >1.330000</td>\n",
       "      <td id=\"T_cdcd0_row9_col2\" class=\"data row9 col2\" >1536192</td>\n",
       "      <td id=\"T_cdcd0_row9_col3\" class=\"data row9 col3\" >0.290000</td>\n",
       "      <td id=\"T_cdcd0_row9_col4\" class=\"data row9 col4\" >2983316</td>\n",
       "      <td id=\"T_cdcd0_row9_col5\" class=\"data row9 col5\" >18.240000</td>\n",
       "      <td id=\"T_cdcd0_row9_col6\" class=\"data row9 col6\" >19</td>\n",
       "      <td id=\"T_cdcd0_row9_col7\" class=\"data row9 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row9_col8\" class=\"data row9 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row9_col9\" class=\"data row9 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row9_col10\" class=\"data row9 col10\" >d__Archaea;p__Methanobacteriota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanobacterium_A;s__</td>\n",
       "      <td id=\"T_cdcd0_row9_col11\" class=\"data row9 col11\" >0.916480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row10\" class=\"row_heading level0 row10\" >C2A3_P_semibin.1662</th>\n",
       "      <td id=\"T_cdcd0_row10_col0\" class=\"data row10 col0\" >96.390000</td>\n",
       "      <td id=\"T_cdcd0_row10_col1\" class=\"data row10 col1\" >0.920000</td>\n",
       "      <td id=\"T_cdcd0_row10_col2\" class=\"data row10 col2\" >451218</td>\n",
       "      <td id=\"T_cdcd0_row10_col3\" class=\"data row10 col3\" >0.350000</td>\n",
       "      <td id=\"T_cdcd0_row10_col4\" class=\"data row10 col4\" >3486530</td>\n",
       "      <td id=\"T_cdcd0_row10_col5\" class=\"data row10 col5\" >11.670000</td>\n",
       "      <td id=\"T_cdcd0_row10_col6\" class=\"data row10 col6\" >83</td>\n",
       "      <td id=\"T_cdcd0_row10_col7\" class=\"data row10 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row10_col8\" class=\"data row10 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row10_col9\" class=\"data row10 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row10_col10\" class=\"data row10 col10\" >d__Archaea;p__Methanobacteriota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanobacterium_A;s__</td>\n",
       "      <td id=\"T_cdcd0_row10_col11\" class=\"data row10 col11\" >0.922030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row11\" class=\"row_heading level0 row11\" >C3D5_F_metabat.427</th>\n",
       "      <td id=\"T_cdcd0_row11_col0\" class=\"data row11 col0\" >96.580000</td>\n",
       "      <td id=\"T_cdcd0_row11_col1\" class=\"data row11 col1\" >2.510000</td>\n",
       "      <td id=\"T_cdcd0_row11_col2\" class=\"data row11 col2\" >238240</td>\n",
       "      <td id=\"T_cdcd0_row11_col3\" class=\"data row11 col3\" >0.400000</td>\n",
       "      <td id=\"T_cdcd0_row11_col4\" class=\"data row11 col4\" >1441087</td>\n",
       "      <td id=\"T_cdcd0_row11_col5\" class=\"data row11 col5\" >12.980000</td>\n",
       "      <td id=\"T_cdcd0_row11_col6\" class=\"data row11 col6\" >14</td>\n",
       "      <td id=\"T_cdcd0_row11_col7\" class=\"data row11 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row11_col8\" class=\"data row11 col8\" >True</td>\n",
       "      <td id=\"T_cdcd0_row11_col9\" class=\"data row11 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row11_col10\" class=\"data row11 col10\" >d__Archaea;p__Thermoproteota;c__Nitrososphaeria;o__Nitrososphaerales;f__JACAEJ01;g__JACAEJ01;s__</td>\n",
       "      <td id=\"T_cdcd0_row11_col11\" class=\"data row11 col11\" >0.937570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cdcd0_level0_row12\" class=\"row_heading level0 row12\" >C3D5_P_metabat.822</th>\n",
       "      <td id=\"T_cdcd0_row12_col0\" class=\"data row12 col0\" >97.090000</td>\n",
       "      <td id=\"T_cdcd0_row12_col1\" class=\"data row12 col1\" >3.590000</td>\n",
       "      <td id=\"T_cdcd0_row12_col2\" class=\"data row12 col2\" >347784</td>\n",
       "      <td id=\"T_cdcd0_row12_col3\" class=\"data row12 col3\" >0.400000</td>\n",
       "      <td id=\"T_cdcd0_row12_col4\" class=\"data row12 col4\" >1813713</td>\n",
       "      <td id=\"T_cdcd0_row12_col5\" class=\"data row12 col5\" >47.840000</td>\n",
       "      <td id=\"T_cdcd0_row12_col6\" class=\"data row12 col6\" >8</td>\n",
       "      <td id=\"T_cdcd0_row12_col7\" class=\"data row12 col7\" >True</td>\n",
       "      <td id=\"T_cdcd0_row12_col8\" class=\"data row12 col8\" >False</td>\n",
       "      <td id=\"T_cdcd0_row12_col9\" class=\"data row12 col9\" >True</td>\n",
       "      <td id=\"T_cdcd0_row12_col10\" class=\"data row12 col10\" >d__Archaea;p__Thermoproteota;c__Nitrososphaeria;o__Nitrososphaerales;f__JACAEJ01;g__JACAEJ01;s__</td>\n",
       "      <td id=\"T_cdcd0_row12_col11\" class=\"data row12 col11\" >0.926520</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7f4221d61700>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcrA_genomes_hq = [g for g in mcrA_genomes if g in summary_hq.index.tolist()]\n",
    "summary_hq.loc[mcrA_genomes_hq, :].style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "75e4469d-f0f0-46f0-b862-0f39d8e8449e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_duplicates(input_list):\n",
    "    seen = set()\n",
    "    return [x for x in input_list if x not in seen and not seen.add(x)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e5262647-d5a2-43aa-a907-a8b9d9103927",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "171cfa5525da4c3d97d49df1a4d1196b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(Output(), VBox(), VBox(children=(Text(value='', description='Filter KO list', pl…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class KOHeatmapDashboard:\n",
    "    def __init__(self, ko_list_path=\"../../input_folder/ko_list\", ko_synonyms_path=\"../../input_folder/ko_synonyms\"):\n",
    "        # Load KO list and synonyms\n",
    "        self.ko_list_df = self.load_ko_list(ko_list_path, ko_synonyms_path)\n",
    "        self.ko_list = (self.ko_list_df.index + \": \" + self.ko_list_df[\"gene_name\"]).to_dict()\n",
    "        self.ko_formatted_names = [\n",
    "            i for i in self.ko_list_df.index + \": \" + self.ko_list_df[\"gene_name\"]\n",
    "        ]\n",
    "\n",
    "        # Custom color scheme\n",
    "        self.colors = [\"#e8e8e8\", \"#df0024\", \"#0085c7\"]\n",
    "\n",
    "        # Global matrix for persistence\n",
    "        self.df_count_ko = pd.DataFrame()\n",
    "\n",
    "        # Widgets\n",
    "        self.filter_text = widgets.Text(\n",
    "            value=\"\",\n",
    "            description=\"Filter KO list\",\n",
    "            placeholder=\"Type to filter...\"\n",
    "        )\n",
    "        self.ko_formatted_names_w = widgets.Select(\n",
    "            options=self.ko_formatted_names,\n",
    "            description=\"KO list\",\n",
    "            rows=40,\n",
    "            layout=widgets.Layout(width='450px')\n",
    "        )\n",
    "        self.genomes_input = widgets.TagsInput(\n",
    "            value=[],\n",
    "            description=\"Genomes\",\n",
    "            allow_duplicates=False\n",
    "        )\n",
    "        self.ko_input = widgets.TagsInput(\n",
    "            value=[],\n",
    "            description=\"KEGG KOs\",\n",
    "            allow_duplicates=False\n",
    "        )\n",
    "        self.heatmap_output = widgets.Output()\n",
    "        self.reset_button = widgets.Button(description=\"Reset Heatmap\")\n",
    "        \n",
    "        # Attach behavior\n",
    "        self.genomes_input.observe(self.update_heatmap, names=\"value\")\n",
    "        self.ko_input.observe(self.update_heatmap, names=\"value\")\n",
    "        self.filter_text.observe(self.filter_ko_list, names=\"value\")\n",
    "        self.ko_formatted_names_w.observe(self.add_to_ko_input, names=\"value\")\n",
    "        self.ko_formatted_names_w.value = None\n",
    "        self.reset_button.on_click(self.reset_table_and_replot)\n",
    "        self.updating_heatmap = False  # Add a flag to prevent duplicate updates\n",
    "\n",
    "    def load_ko_list(self, ko_list_path, ko_synonyms_path):\n",
    "        \"\"\"Load and format the KO list with synonyms.\"\"\"\n",
    "        ko_list_df = pd.read_csv(ko_list_path, sep=\"\\t\", index_col=0)\n",
    "        ko_list_df = ko_list_df.loc[:, [\"definition\"]]\n",
    "        ko_list_df.columns = [\"gene_name\"]\n",
    "\n",
    "        ko_synonyms = pd.read_csv(ko_synonyms_path, sep=\"\\t\", index_col=0, header=None, names=[\"synonyms\"])\n",
    "        ko_synonyms.fillna(\"\", inplace=True)\n",
    "\n",
    "        ko_list_df = ko_list_df.merge(ko_synonyms, left_index=True, right_index=True)\n",
    "        ko_list_df[\"gene_name\"] = ko_list_df[\"synonyms\"] + \", \" + ko_list_df[\"gene_name\"]\n",
    "        ko_list_df.drop(columns=[\"synonyms\"], inplace=True)\n",
    "        ko_list_df.index.name = \"KO\"\n",
    "        return ko_list_df\n",
    "\n",
    "    def read_df(self, sample):\n",
    "        \"\"\"Optimized function to read sample data.\"\"\"\n",
    "        return pd.read_csv(f\"../../input_folder/dram_faa/{sample}/annotations.tsv\", sep=\"\\t\")\n",
    "\n",
    "    def count_sample_ko(self, annot_df, ko):\n",
    "        \"\"\"Optimized function to count KO occurrences.\"\"\"\n",
    "        return annot_df[\"ko_id\"].eq(ko).sum()\n",
    "\n",
    "    def update_matrix(self, genomes, genes):\n",
    "        \"\"\"Update the KO-genome occurrence matrix.\"\"\"\n",
    "        new_columns = [self.ko_list[gene] for gene in genes]\n",
    "        self.df_count_ko = self.df_count_ko.reindex(columns=new_columns)\n",
    "\n",
    "        for gene in genes:\n",
    "            if self.ko_list[gene] not in self.df_count_ko.columns:\n",
    "                self.df_count_ko[self.ko_list[gene]] = 0\n",
    "\n",
    "        for genome in genomes:\n",
    "            if genome not in self.df_count_ko.index:\n",
    "                self.df_count_ko.loc[genome] = 0\n",
    "\n",
    "        for genome in genomes:\n",
    "            for gene in genes:\n",
    "                gene_name = self.ko_list[gene]\n",
    "                counts = self.count_sample_ko(self.read_df(genome), gene)\n",
    "                self.df_count_ko.at[genome, gene_name] = counts\n",
    "\n",
    "    def update_heatmap(self, change=None):\n",
    "        \"\"\"Update the heatmap visualization.\"\"\"\n",
    "        if self.updating_heatmap:\n",
    "            return  # Skip if already updating\n",
    "    \n",
    "        self.updating_heatmap = True  # Set the flag\n",
    "        with self.heatmap_output:\n",
    "            clear_output(wait=True)\n",
    "    \n",
    "            # Process genomes_input to handle whitespace-separated values\n",
    "            processed_genomes = []\n",
    "            for genome in self.genomes_input.value:\n",
    "                genome = re.sub(r\"[\\'\\[\\],]\", \" \", genome)\n",
    "                if \" \" in genome:  # Check for whitespace\n",
    "                    processed_genomes.extend(genome.split())  # Split and add the items in order\n",
    "                else:\n",
    "                    processed_genomes.append(genome)\n",
    "    \n",
    "            self.genomes_input.value = processed_genomes  # Update the value without altering order\n",
    "            selected_genomes = self.genomes_input.value\n",
    "    \n",
    "            processed_kos = []\n",
    "            for ko in self.ko_input.value:\n",
    "                ko = re.sub(r\"[\\+\\-,()]\", \" \", ko)\n",
    "                if \" \" in ko:\n",
    "                    kos = [k for k in ko.split() if re.match(\"^K[0-9]*\", k)]\n",
    "                    kos = remove_duplicates(kos)\n",
    "                    processed_kos.extend(kos)\n",
    "                else:\n",
    "                    processed_kos.append(ko)\n",
    "    \n",
    "            self.ko_input.value = processed_kos\n",
    "            selected_kos = self.ko_input.value\n",
    "    \n",
    "            self.update_matrix(selected_genomes, selected_kos)\n",
    "    \n",
    "            mk_max = self.df_count_ko.max().max()\n",
    "            mk_val1 = 1 / mk_max if mk_max > 0 else 1\n",
    "    \n",
    "            num_rows, num_cols = self.df_count_ko.shape\n",
    "            cell_size = 30  # Size per cell in pixels\n",
    "            min_width = 400  # Minimum width for heatmap\n",
    "            min_height = 300  # Minimum height for heatmap\n",
    "    \n",
    "            # Dynamically adjust the figure dimensions\n",
    "            heatmap_width = max(num_cols * cell_size, min_width)\n",
    "            heatmap_height = max(num_rows * cell_size, min_height)\n",
    "            total_width = heatmap_width + 200  # Add 200 px for y-axis tick labels\n",
    "            total_height = heatmap_height + 200  # Add 200 px for x-axis tick labels\n",
    "    \n",
    "            fig = px.imshow(self.df_count_ko, aspect=\"auto\", text_auto=True)\n",
    "            fig.update_layout(\n",
    "                width=total_width,\n",
    "                height=total_height,\n",
    "                font=dict(size=18),\n",
    "                margin=dict(l=200, t=50, r=50, b=200),  # Space for labels and a clean layout\n",
    "            )\n",
    "    \n",
    "            tickvals = list(range(self.df_count_ko.shape[1]))\n",
    "            xticktext = [re.sub(\",.*\", \"\", c) for c in self.df_count_ko.columns.tolist()]\n",
    "            fig.update_xaxes(tickangle=90, ticktext=xticktext, tickvals=tickvals)\n",
    "    \n",
    "            fig.update_traces(xgap=1, ygap=1)\n",
    "    \n",
    "            colorscale = [\n",
    "                (0.0, self.colors[0]),\n",
    "                (mk_val1 - 0.25 * mk_val1, self.colors[0]),\n",
    "                (mk_val1, self.colors[1]),\n",
    "                (mk_val1 + 0.25 * mk_val1, self.colors[2]),\n",
    "                (1, self.colors[2])\n",
    "            ] if mk_val1 < 1 else [\n",
    "                (0.0, self.colors[0]),\n",
    "                (0.75, self.colors[1]),\n",
    "                (1, self.colors[2])\n",
    "            ]\n",
    "            fig.update_coloraxes(colorscale=colorscale)\n",
    "            fig.show(config={\"editable\": True})\n",
    "    \n",
    "        self.updating_heatmap = False  # Reset the flag\n",
    "\n",
    "    def reset_table_and_replot(self, change=None):\n",
    "        \"\"\"Reset the matrix and refresh the heatmap.\"\"\"\n",
    "        self.df_count_ko = pd.DataFrame()\n",
    "        selected_genomes = self.genomes_input.value\n",
    "        selected_kos = self.ko_input.value\n",
    "        new_columns = [self.ko_list[gene] for gene in selected_kos]\n",
    "        self.df_count_ko = self.df_count_ko.reindex(columns=new_columns)\n",
    "        self.update_matrix(selected_genomes, selected_kos)\n",
    "        self.update_heatmap(None)\n",
    "\n",
    "    def filter_ko_list(self, change):\n",
    "        \"\"\"Filter KO list based on user input.\"\"\"\n",
    "        filter_value = self.filter_text.value.lower()\n",
    "        filtered_options = [\n",
    "            ko for ko in self.ko_formatted_names if filter_value in ko.lower()\n",
    "        ]\n",
    "        self.ko_formatted_names_w.options = filtered_options\n",
    "\n",
    "        if not filter_value:\n",
    "            self.ko_formatted_names_w.value = None\n",
    "\n",
    "    def add_to_ko_input(self, change):\n",
    "        \"\"\"Add selected KO to the input list.\"\"\"\n",
    "        selected_item = self.ko_formatted_names_w.value\n",
    "        if selected_item:\n",
    "            selected_ko = selected_item.split(\":\")[0]\n",
    "            if selected_ko not in self.ko_input.value:\n",
    "                self.ko_input.value = self.ko_input.value + [selected_ko]\n",
    "\n",
    "    def display(self):\n",
    "        \"\"\"Display the dashboard.\"\"\"\n",
    "        layout = widgets.VBox([\n",
    "            widgets.HBox([\n",
    "                self.heatmap_output,\n",
    "                widgets.VBox(),\n",
    "                widgets.VBox([self.filter_text, self.ko_formatted_names_w, self.reset_button])\n",
    "            ]),\n",
    "            widgets.HTML(value=\"<b>Selected Genomes </b>\"),\n",
    "            self.genomes_input,\n",
    "            widgets.HTML(value=\"<b>Selected KEGG KO</b>\"),\n",
    "            self.ko_input\n",
    "        ])\n",
    "        display(layout)\n",
    "\n",
    "\n",
    "# Usage\n",
    "dashboard = KOHeatmapDashboard()\n",
    "dashboard.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "eb7a35fb-0ccd-4cbf-b609-ec7c63107e63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ce4f631b07d04d958fd2647265a33758",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(VBox(children=(Text(value='', description='Filter modules', placeholder='Type to filter...'), S…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "from Bio.KEGG import REST\n",
    "import pandas as pd\n",
    "\n",
    "class KEGGModuleDashboard:\n",
    "    def __init__(self):\n",
    "        # Step 1: Fetch KEGG modules and prepare data\n",
    "        self.module_df = self.fetch_kegg_modules()\n",
    "        self.modules_formatted_names = (\n",
    "            self.module_df[\"module\"] + \": \" + self.module_df[\"name\"]\n",
    "        ).tolist()\n",
    "\n",
    "        # Step 2: Define widgets\n",
    "        self.filter_text = widgets.Text(\n",
    "            value=\"\",\n",
    "            description=\"Filter modules\",\n",
    "            placeholder=\"Type to filter...\"\n",
    "        )\n",
    "\n",
    "        self.modules_select_w = widgets.Select(\n",
    "            options=self.modules_formatted_names,\n",
    "            description=\"Modules list\",\n",
    "            rows=20,\n",
    "            layout=widgets.Layout(width=\"450px\")\n",
    "        )\n",
    "\n",
    "        self.module_info_output = widgets.Output()\n",
    "\n",
    "        # Step 3: Attach widget behaviors\n",
    "        self.filter_text.observe(self.filter_modules, names=\"value\")\n",
    "        self.modules_select_w.observe(self.display_module_info, names=\"value\")\n",
    "\n",
    "        # Step 4: Create the dashboard layout\n",
    "        self.dashboard = widgets.HBox(\n",
    "            [widgets.VBox([self.filter_text, self.modules_select_w]), self.module_info_output]\n",
    "        )\n",
    "\n",
    "    def fetch_kegg_modules(self):\n",
    "        \"\"\"Fetch the list of KEGG modules and return as a DataFrame.\"\"\"\n",
    "        module_list_data = REST.kegg_list(\"module\").read()\n",
    "        return pd.DataFrame(\n",
    "            [line.split(\"\\t\") for line in module_list_data.strip().split(\"\\n\")],\n",
    "            columns=[\"module\", \"name\"]\n",
    "        )\n",
    "\n",
    "    def filter_modules(self, change):\n",
    "        \"\"\"Update the options in the module select widget based on filter text.\"\"\"\n",
    "        search_text = change[\"new\"]\n",
    "        if search_text == \"\":\n",
    "            filtered_modules = self.modules_formatted_names\n",
    "        else:\n",
    "            filtered_modules = [\n",
    "                name for name in self.modules_formatted_names\n",
    "                if search_text.lower() in name.lower()\n",
    "            ]\n",
    "        self.modules_select_w.options = filtered_modules\n",
    "\n",
    "    def display_module_info(self, change):\n",
    "        \"\"\"Fetch and display module information for the selected module.\"\"\"\n",
    "        self.module_info_output.clear_output()\n",
    "        selected_module = change[\"new\"].split(\":\")[0]  # Extract module ID\n",
    "        if selected_module:\n",
    "            with self.module_info_output:\n",
    "                module_info = REST.kegg_get(f\"module:{selected_module}\").readlines()\n",
    "                for l in module_info:\n",
    "                    if \"NAME\" in l:\n",
    "                        name = l\n",
    "                        break\n",
    "                name = re.sub(\"NAME *\", \"\", name)\n",
    "                print(f\"{selected_module}: {name}\")\n",
    "                module_info_description = \"\".join(module_info)\n",
    "                print(module_info_description)\n",
    "\n",
    "    def display(self):\n",
    "        \"\"\"Display the dashboard.\"\"\"\n",
    "        display(self.dashboard)\n",
    "\n",
    "# Create and display the KEGG Module Dashboard\n",
    "kegg_dashboard = KEGGModuleDashboard()\n",
    "kegg_dashboard.display()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c36dc54-6c1f-43be-aa39-8076ef7b2f1a",
   "metadata": {},
   "source": [
    "# KIV below"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ea63d18e-44ca-4005-872f-f77880ae2923",
   "metadata": {},
   "source": [
    "\n",
    "# Load KO list dataframe\n",
    "# Load KO list dataframe\n",
    "ko_list_df_path = \"ko_list\"\n",
    "ko_list_df = pd.read_csv(ko_list_df_path, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# Clean KO list and format names\n",
    "ko_list_df = ko_list_df.loc[:, [\"definition\"]]\n",
    "ko_list_df.columns = [\"gene_name\"]\n",
    "ko_synonyms = pd.read_csv(\"ko_synonyms2\", sep=\"\\t\", index_col=0, header=None, names=[\"synonyms\"])\n",
    "ko_synonyms.fillna(\"\", inplace=True)\n",
    "\n",
    "ko_list_df = ko_list_df.merge(ko_synonyms, left_index=True, right_index=True)\n",
    "ko_list_df[\"gene_name\"] = ko_list_df[\"synonyms\"] + \", \" + ko_list_df[\"gene_name\"]\n",
    "ko_list_df.drop(columns=[\"synonyms\"], inplace=True)\n",
    "ko_list_df.index.name = \"KO\"\n",
    "ko_list = (ko_list_df.index + \": \" + ko_list_df[\"gene_name\"]).to_dict()\n",
    "\n",
    "max_char = 50\n",
    "ko_formatted_names = [\n",
    "     i for i in ko_list_df.index + \": \" + ko_list_df[\"gene_name\"]\n",
    "]\n",
    "# f\"{i[:max_char]}..\" if len(i) > max_char else i for i in ko_list_df.index + \": \" + ko_list_df[\"gene_name\"]\n",
    "\n",
    "ko_formatted_names_w = widgets.Select(\n",
    "    options=ko_formatted_names,\n",
    "    description=\"KO list\",\n",
    "    rows=40,\n",
    "    layout=widgets.Layout(width='450px')\n",
    ")\n",
    "\n",
    "# Optimized function to read sample data\n",
    "def read_df(sample):\n",
    "    return pd.read_csv(f\"./dram_faa/{sample}/annotations.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Optimized function to count KO occurrences\n",
    "def count_sample_ko(annot_df, ko):\n",
    "    return annot_df[\"ko_id\"].eq(ko).sum()\n",
    "\n",
    "# Optimize matrix creation by avoiding repeated file reads\n",
    "def matrix_ko(genome_selection, gene_selection):\n",
    "    df_count_ko = pd.DataFrame(index=genome_selection)\n",
    "\n",
    "    for gene in gene_selection:\n",
    "        # gene_name = ko_list[gene][:max_char] + \"...\" if len(ko_list[gene]) > max_char else ko_list[gene]\n",
    "        gene_name = ko_list[gene]\n",
    "        counts = {genome: count_sample_ko(read_df(genome), gene) for genome in genome_selection}\n",
    "        df_count_ko[gene_name] = df_count_ko.index.map(counts)\n",
    "\n",
    "    return df_count_ko\n",
    "\n",
    "# Load genomes list once\n",
    "# with open(\"genomes.txt\", \"r\") as f:\n",
    "#     genomes = f.read().split()\n",
    "\n",
    "\n",
    "init_genomes = genomes[:6]\n",
    "genomes_input = widgets.TagsInput(\n",
    "    value=init_genomes,\n",
    "    description=\"Give some genomes names\",\n",
    "    allow_duplicates=False\n",
    ")\n",
    "\n",
    "init_ko_list = [\"K00370\", \"K00371\", \"K00374\", \"K00373\"]\n",
    "ko_input = widgets.TagsInput(\n",
    "    value=init_ko_list,\n",
    "    description=\"Give some KEGG KOs\",\n",
    "    allow_duplicates=False\n",
    ")\n",
    "\n",
    "filter_text = widgets.Text(\n",
    "    value=\"\",\n",
    "    description=\"Filter KO list\",\n",
    "    placeholder=\"Type to filter...\"\n",
    ")\n",
    "\n",
    "# Heatmap output widget\n",
    "heatmap_output = widgets.Output()\n",
    "\n",
    "# Custom color scheme\n",
    "colors = [\"#e8e8e8\", \"#df0024\", \"#0085c7\"]\n",
    "\n",
    "# Update heatmap function\n",
    "def update_heatmap(change):\n",
    "    with heatmap_output:\n",
    "        clear_output(wait=True)\n",
    "\n",
    "        selected_genomes = genomes_input.value\n",
    "        selected_kos = ko_input.value\n",
    "\n",
    "        # Generate the matrix\n",
    "        mk = matrix_ko(selected_genomes, selected_kos)\n",
    "        mk_max = mk.max().max()\n",
    "        mk_val1 = 1 / mk_max if mk_max > 0 else 1\n",
    "\n",
    "        # Plot the heatmap\n",
    "        fig = px.imshow(mk, aspect=\"auto\", text_auto=True)\n",
    "        height = 1000\n",
    "        width = 1500\n",
    "        if len(ko_input.value) * 80 < width:\n",
    "            width = len(ko_input.value) * 80\n",
    "        fig.update_layout(height=height, width=width, yaxis_scaleanchor=\"x\")\n",
    "        # fig.update_layout(yaxis = dict(scaleanchor = 'x'))\n",
    "        \n",
    "        tickvals = list(range(mk.shape[1]))\n",
    "        xticktext = [re.sub(\",.*\", \"\", c) for c in mk.columns.tolist()]\n",
    "\n",
    "        \n",
    "        fig.update_xaxes(tickangle=90, ticktext=xticktext, tickvals=tickvals, ticklen=200)\n",
    "# fig.update_xaxes(\n",
    "#     tickvals=ticks_locations,\n",
    "#     ticktext=ticks,\n",
    "#     tickcolor='black',\n",
    "#     ticks='outside',\n",
    "#     ticklabelposition='outside right',\n",
    "#     # tickson='boundaries',\n",
    "# )\n",
    "\n",
    "        colorscale = [\n",
    "            (0.0, colors[0]),\n",
    "            (mk_val1 - 0.25 * mk_val1, colors[0]),\n",
    "            (mk_val1, colors[1]),\n",
    "            (mk_val1 + 0.25 * mk_val1, colors[2]),\n",
    "            (1, colors[2])\n",
    "        ] if mk_val1 < 1 else [\n",
    "            (0.0, colors[0]),\n",
    "            (0.75, colors[1]),\n",
    "            (1, colors[2])\n",
    "        ]\n",
    "\n",
    "        fig.update_coloraxes(colorscale=colorscale)\n",
    "        fig.show(config={\"editable\": True})\n",
    "\n",
    "# Register update function\n",
    "genomes_input.observe(update_heatmap, names='value')\n",
    "ko_input.observe(update_heatmap, names='value')\n",
    "\n",
    "# Filter KO list based on user input\n",
    "# def filter_ko_list(change):\n",
    "#     filter_value = filter_text.value.lower()\n",
    "#     ko_formatted_names_w.options = [\n",
    "#         ko for ko in ko_formatted_names if filter_value in ko.lower()\n",
    "#     ]\n",
    "\n",
    "def filter_ko_list(change):\n",
    "    ko_formatted_names_w.value = None\n",
    "    filter_value = filter_text.value.lower()\n",
    "    filtered_options = [\n",
    "        ko for ko in ko_formatted_names if filter_value in ko.lower()\n",
    "    ]\n",
    "    ko_formatted_names_w.options = filtered_options\n",
    "\n",
    "\n",
    "filter_text.observe(filter_ko_list, names='value')\n",
    "\n",
    "# Add selected KO to input list\n",
    "def add_to_ko_input(change):\n",
    "    selected_item = ko_formatted_names_w.value\n",
    "    if selected_item:\n",
    "        selected_ko = selected_item.split(\":\")[0]\n",
    "        if selected_ko not in ko_input.value:\n",
    "            ko_input.value = ko_input.value + [selected_ko]\n",
    "\n",
    "ko_formatted_names_w.observe(add_to_ko_input, names='value')\n",
    "\n",
    "# Initial heatmap\n",
    "update_heatmap(None)\n",
    "\n",
    "# Display the widgets and the interactive output\n",
    "widgets.VBox([\n",
    "    widgets.HBox([heatmap_output, widgets.VBox(), widgets.VBox([filter_text, ko_formatted_names_w])]),\n",
    "    widgets.HTML(value=\"<b>Selected Genomes </b>\"),\n",
    "    genomes_input,\n",
    "    widgets.HTML(value=\"<b>Selected KEGG KO</b>\"),\n",
    "    ko_input\n",
    "])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf4fc189-9729-4b14-b04b-d296a7007bf5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
